### PR TITLE
fix(#2440): Make dovecot sieve directories writeable

### LINKED ISSUE #2440: The LDA Sieve plugin does not have permission to save global Sieve script binaries
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

Folder permission error compiling Sieve global scripts

### What happened and when does this occur?

```Markdown
The file `/usr/lib/dovecot/sieve-global/before/50-before.dovecot.svbin` is not created due to permissions issues
```

### What did you expect to happen?

```Markdown
No Sieve errors compiling scripts
```

### How do we replicate the issue?

1. created file `config/dovecot.cf`:
```conf
plugin {
    #https://github.com/docker-mailserver/docker-mailserver/issues/2277#issuecomment-963589913
    #sieve_global_extensions = +editheader
    sieve_extensions = +editheader
}
```

2. created file `config/before.dovecot.sieve`:
```sieve
require [
    "editheader",
    "fileinto",
    "mailbox",
    "variables"
];

#https://answers.microsoft.com/it-it/outlook_com/forum/all/mail-di-spam-dal-dominio-onmicrosoftcom/2ff8b433-576f-493f-9a54-7ad34a092457
if allof (
    header :contains "Received-SPF" "outbound.protection.outlook.com",
    header :is "X-MS-Exchange-CrossTenant-FromEntityHeader" "Hosted",
    header :is "To" "Undisclosed recipients:;"
) {

    # store the original subject in a variable that later rules can use
    if header :matches "Subject" "*" {
        set "subject" "${1}";
    }

    if not header "X-Spam-Flag" "YES" {
        deleteheader "X-Spam-Flag";
        addheader "X-Spam-Flag" "YES";
    }

    if not header :contains "Subject" "***SPAM*****" {
        deleteheader "Subject";
        addheader :last "Subject" "***SPAM*** ${subject}";
    }

    fileinto "Junk";
}
```

3. restarted DMS

### DMS version

v10.4.0

### What operating system is DMS running on?

Linux

### What instruction set architecture is DMS running on?

x86_64 / AMD64

### What container orchestration tool are you using?

Docker Compose

### docker-compose.yml

```yaml
version: '3.9'

services:
  mailserver:
    image: docker.io/mailserver/docker-mailserver:10.4.0
    container_name: mailserver
    hostname: my.hostname.tld
    env_file: .env
    ports:
      - "25:25"
      - "110:110"
      - "143:143"
      - "465:465"
      - "587:587"
      - "993:993"
      - "995:995"
    volumes:
      - ${PWD}/sys/mailserver/data:/var/mail
      - ${PWD}/sys/mailserver/state:/var/mail-state
      - ${PWD}/sys/mailserver/logs:/var/log/mail
      - ${PWD}/sys/mailserver/config:/tmp/docker-mailserver
      - ${PWD}/sys/certbot/certs:/etc/letsencrypt
      - /etc/localtime:/etc/localtime:ro
    restart: always
    stop_grace_period: 1m
    cap_add:
      - NET_ADMIN
      - SYS_PTRACE
```

### Relevant log output

Startup error logs:

```sh
$ docker-compose logs -f
mailserver    | 50-before.dovecot: line 13: error: require command: unknown Sieve capability `editheader'.
mailserver    | 50-before.dovecot: line 38: error: unknown command 'deleteheader' (only reported once at first occurrence).
mailserver    | 50-before.dovecot: line 39: error: unknown command 'addheader' (only reported once at first occurrence).
mailserver    | 50-before.dovecot: error: validation failed.
mailserver    | sievec: Fatal: failed to compile sieve script '/usr/lib/dovecot/sieve-global/before/50-before.dovecot.sieve'
```

Log upon receipt of an email:
```sh
$ docker exec -it mailserver setup debug show-mail-logs
Error: sieve: binary /usr/lib/dovecot/sieve-global/before/50-before.dovecot.svbin: save: failed to create temporary file: open(/usr/lib/dovecot/sieve-global/before/50-before.dovecot.svbin.) failed: Permission denied (euid=5000(docker) egid=5000(docker) missing +w perm: /usr/lib/dovecot/sieve-global/before, dir owner missing perms)
Error: sieve: The LDA Sieve plugin does not have permission to save global Sieve script binaries; global Sieve scripts like `/usr/lib/dovecot/sieve-global/before/50-before.dovecot.sieve' need to be pre-compiled using the sievec tool
```

### Other relevant information

To solve the problem I added this to `user-patches.sh`:

```sh
chmod +w /usr/lib/dovecot/sieve-global;
chmod +w /usr/lib/dovecot/sieve-global/after;
chmod +w /usr/lib/dovecot/sieve-global/before;
```

This ensures that the Sieve global filters run correctly, despite the fact that some errors log remains on startup (_50-before.dovecot: line 13: error: require command: unknown Sieve capability `editheader'_ ...)

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [ ] I am inexperienced with mail servers
- [ ] I am uncomfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

### Improvements to this form?

_No response_

### PR DESCRIPTION
**Pull request by casperklein**
# Description

This PR fixes #2440

The correct permissions are set here: https://github.com/docker-mailserver/docker-mailserver/blob/dca2372431f8040485f1162365589d9387b13a93/Dockerfile#L118

But for some reason, the write permission is removed later: https://github.com/docker-mailserver/docker-mailserver/blob/dca2372431f8040485f1162365589d9387b13a93/target/scripts/startup/setup-stack.sh#L262

This PR makes the three directories writeable again.

Disclaimer: I don't use sieve myself. So better someone more familiar with sieve reviews this 😉 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by polarathene** [comment]
> Disclaimer: I don't use sieve myself. So better someone more familiar with sieve reviews this wink

I don't either, but both lines were introduced into the code base [via the same PR](https://github.com/docker-mailserver/docker-mailserver/pull/934/files#diff-6d88c6c2abe074135b8efac1cf0dca97dd288a71720cdcd6ac48c2a6f85dcf08R502) in 2018 for v6 release.

There is no commit message or discussion regarding the difference in permissions. Nor in linked issues or the original PR (_the `chmod 550 ..` line doesn't exist in that one_) it referenced.

File: target/scripts/startup/setup-stack.sh
Comment (polarathene) [inline target/scripts/startup/setup-stack.sh:262]: Since this is covered by the `Dockerfile` already, is there any need for this line here too? Or is it due to changes since at runtime that we or user adds?

I guess there's no harm in doing so, I don't mind either way. I had a quick search for `dovecot sieve chmod` on the web and nothing seemed to stand out, it's unclear why root was being made read-only here by original author.

**Review by polarathene — APPROVED** [review]
I'm not sure how important the test is, this is not a case of accidentally introducing a regression change, it appears to have always been this way since it was introduced. `git blame` will at least correctly provide context via this PR for any future maintenance (_except the `Dockerfile`, but grepping the repo for extra context should assist with that_).

I don't mind the tests, I'm just not sure how useful they are at present.

File: target/scripts/startup/setup-stack.sh
Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:262]: > Since this is covered by the Dockerfile already, is there any need for this line here too? Or is it due to changes since at runtime that we or user adds?

Good point. Maybe just removing the `chmod` line is enough.

@17Halbe Can you explain, what your intention was?

Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:262]: > I don't mind the tests, I'm just not sure how useful they are at present.

I've added them to prevent future regression made by any PR.

Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:262]: I understand why you added them, but in this case a regression was never introduced, the behaviour is the same since it was contributed. This wasn't different before and missed in a review by maintainers due to passing tests (_other than the original PR itself where it wasn't questioned/noticed_).

I don't think we have permissions check tests where we do this elsewhere as well? It's fine if you think there is value in it still though.

---

I wouldn't expect @17Halbe to recall such a detail from almost 4 years ago :sweat_smile: but maybe! :crossed_fingers:

**Review by georglauterbach — APPROVED** [review]

**Comment by casperklein** [comment]
I would like to wait for [feedback](https://github.com/docker-mailserver/docker-mailserver/issues/2440#issuecomment-1053712738), before merging this.

**Comment by polarathene** [comment]
@casperklein did you want this PR to be closed like that?

**Comment by casperklein** [comment]
Not sure TBH. We can reopen this (for a while, but not forever 😆), to wait if someone comes up with a proper solution. See also https://github.com/docker-mailserver/docker-mailserver/issues/2440#issuecomment-1073347429

**Comment by georglauterbach** [comment]
> Not sure TBH. We can reopen this (for a while, but not forever laughing), to wait if someone comes up with a proper solution. See also [#2440 (comment)](https://github.com/docker-mailserver/docker-mailserver/issues/2440#issuecomment-1073347429)

It has been a month now - I'd go for closing this as we already have enough PRs marked as WIP that are not being merged soon :)

**Pull request closed without merging**
